File: src/chain-observers/providers/content-metadata-provider.ts
import { TAccountName } from "@hiveio/wax";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { ContentMetadataClassifier, TContentMetadataAuthorData } from "../classifiers/content-metadata-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

// Common interface for blog content data (posts and comments)
export type TContentMetadataProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: TContentMetadataAuthorData;
};

// Comment metadata provider implementation
export interface ICommentMetadataProviderOptions {
  authors: TAccountName[];
}

// Post metadata provider implementation
export interface IPostMetadataProviderOptions {
  authors: TAccountName[];
}

export interface ICommentMetadataProviderData<TAccounts extends Array<TAccountName>> {
  commentsMetadata: Partial<TContentMetadataProvided<TAccounts>>;
}

export interface IPostMetadataProviderData<TAccounts extends Array<TAccountName>> {
  postsMetadata: Partial<TContentMetadataProvided<TAccounts>>;
}

// Base class for blog content providers (posts and comments)
export abstract class ContentMetadataProvider<
  TStructure extends object = object,
  TAccounts extends Array<TAccountName> = Array<TAccountName>,
  TOptions extends object = object
> extends ProviderBase<TOptions, TStructure> {
  public readonly authors = new Set<TAccountName>();
  protected readonly isPost: boolean;

  public constructor(isPost: boolean) {
    super();
    this.isPost = isPost;
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [ContentMetadataClassifier];
  }

  public abstract pushOptions(options: TOptions): void;

  public async createProviderData(data: TProviderEvaluationContext): Promise<Partial<TContentMetadataProvided<TAccounts>>> {
    const { contentData } = await data.get(ContentMetadataClassifier);

    const result: Partial<TContentMetadataProvided<TAccounts>> = {};
    for (const account of this.authors)
      if (contentData[account])
        for(const permlink in contentData[account]) {
          const postMetadata = contentData[account][permlink];

          const postIndicator = postMetadata.parentAuthor === "";
          if (this.isPost !== postIndicator)
            continue;

          result[account] = result[account] ?? {};
          result[account][permlink] = postMetadata;
        }

    return result;
  }
}

export class CommentMetadataProvider<TAccounts extends Array<TAccountName> = Array<TAccountName>>
  extends ContentMetadataProvider<ICommentMetadataProviderData<TAccounts>, TAccounts, ICommentMetadataProviderOptions> {
  public constructor() {
    super(false); // False = not posts, i.e., comments
  }

  public pushOptions(options: ICommentMetadataProviderOptions): void {
    for (const account of options.authors)
      this.authors.add(account);
  }

  public get baseStructure(): ICommentMetadataProviderData<TAccounts> {
    return {
      commentsMetadata: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<ICommentMetadataProviderData<TAccounts>> {
    const result = this.baseStructure;

    result.commentsMetadata = await this.createProviderData(data);

    return result;
  }
}

export class PostMetadataProvider<TAccounts extends Array<TAccountName> = Array<TAccountName>>
  extends ContentMetadataProvider<IPostMetadataProviderData<TAccounts>, TAccounts, IPostMetadataProviderOptions> {
  public constructor() {
    super(true); // True = posts
  }

  public pushOptions(options: IPostMetadataProviderOptions): void {
    for (const account of options.authors)
      this.authors.add(account);
  }

  public get baseStructure(): IPostMetadataProviderData<TAccounts> {
    return {
      postsMetadata: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IPostMetadataProviderData<TAccounts>> {
    const result = this.baseStructure;

    result.postsMetadata = await this.createProviderData(data);

    return result;
  }
}

File: python/workerbee/chain_observers/providers/content_metadata_provider.py
"""PostMetadataProvider + CommentMetadataProvider — content metadata providers.

Mirrors src/chain-observers/providers/content-metadata-provider.ts.
Separates posts (parent_author == "") from comments (parent_author != "").
"""

from __future__ import annotations

from abc import abstractmethod
from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.content_metadata_classifier import ContentMetadataClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import CommentsMetadataPayload, ContentMetadata, PostsMetadataPayload

    # account -> permlink -> reshaped metadata
    ContentMetadataResult = dict[str, dict[str, ContentMetadata]]

ContentMetadataOptions = dict[str, list[str]]


class ContentMetadataProviderBase(ProviderBase):
    """Base class for post/comment metadata providers."""

    def __init__(self, is_post: bool) -> None:
        self.authors: OrderedSet[str] = OrderedSet()
        self._is_post: bool = is_post

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [ContentMetadataClassifier]

    @abstractmethod
    def push_options(self, options: ContentMetadataOptions) -> None: ...

    async def create_provider_data(self, data: DataEvaluationContext) -> ContentMetadataResult:
        content_result = await data.get(ContentMetadataClassifier)
        content_data: dict[str, dict[str, ContentMetadata]] = content_result.get("content_data", {})

        result: ContentMetadataResult = {}
        for account in self.authors:
            account_data = content_data.get(account)
            if not account_data:
                continue
            for permlink, post_metadata in account_data.items():
                post_indicator = post_metadata.get("parent_author", "") == ""
                if self._is_post != post_indicator:
                    continue

                if account not in result:
                    result[account] = {}
                result[account][permlink] = post_metadata

        return result


class PostMetadataProvider(ContentMetadataProviderBase):
    def __init__(self) -> None:
        super().__init__(is_post=True)

    def push_options(self, options: ContentMetadataOptions) -> None:
        authors = options.get("authors", [])
        if not authors:
            raise ValueError("authors must not be empty")
        for account in authors:
            self.authors.add(account)

    async def provide(self, data: DataEvaluationContext) -> PostsMetadataPayload:
        return {
            "posts_metadata": await self.create_provider_data(data),
        }


class CommentMetadataProvider(ContentMetadataProviderBase):
    def __init__(self) -> None:
        super().__init__(is_post=False)

    def push_options(self, options: ContentMetadataOptions) -> None:
        authors = options.get("authors", [])
        if not authors:
            raise ValueError("authors must not be empty")
        for account in authors:
            self.authors.add(account)

    async def provide(self, data: DataEvaluationContext) -> CommentsMetadataPayload:
        return {
            "comments_metadata": await self.create_provider_data(data),
        }
